src/enzo/CommunicationShareActiveParticles.C

```diff
@@ -0,0 +1,185 @@
+/
+/  PURPOSE: Takes a list of active particles moves and sends/receives
+/           them to all processors
+/
+************************************************************************/
+
+#ifdef USE_MPI
+#endif /* USE_MPI */
+
+#include "preincludes.h"
+#include "ErrorExceptions.h"
+#include "macros_and_parameters.h"
+#include "typedefs.h"
+#include "global_data.h"
+#include "Fluxes.h"
+#include "GridList.h"
+#include "ExternalBoundary.h"
+#include "Grid.h"
+#include "ActiveParticle.h"
+#include "CommunicationUtilities.h"
+#include "SortCompareFunctions.h"
+
+void my_exit(int status);
+
+int CommunicationShareActiveParticles(int *NumberToMove, 
+				      ActiveParticleList<ActiveParticleType> &SendList,
+				      int &NumberOfReceives, 
+				      ActiveParticleList<ActiveParticleType> &SharedList)
+{
+
+  int i, type, proc, ap_id;
+  int SizeOfSends, NumberOfNewParticles;
+  ActiveParticleType_info *ap_info;
+
+  int TotalNumberToMove = 0, GlobalNumberToMove;
+  for (proc = 0; proc < NumberOfProcessors; proc++)
+    TotalNumberToMove += NumberToMove[proc];
+  GlobalNumberToMove = TotalNumberToMove;
+#ifdef USE_MPI
+  CommunicationAllReduceValues(&GlobalNumberToMove, 1, MPI_SUM);
+#endif
+  if (GlobalNumberToMove == 0) return SUCCESS;
+
+#ifdef USE_MPI
+  MPI_Arg Count;
+  MPI_Arg SendCount;
+  MPI_Arg RecvCount;
+  MPI_Arg stat;
+#endif /* USE_MPI */
+
+  if (NumberOfProcessors > 1) {
+
+#ifdef USE_MPI
+
+    /* We will have one collective communication per particle type.
+       In the future, there might be a way to consolidate all of the
+       buffers into one buffer and communicate it as a whole. */
+
+    for (type = 0; type < EnabledActiveParticlesCount; type++) {
+
+      ap_info = EnabledActiveParticles[type];
+      ap_id = ap_info->GetEnabledParticleID();
+
+      /* Get counts from each processor to allocate buffers. */
+
+      int NumberToSend, local_buffer_size, instance_size;
+
+      NumberToSend = 0;
+      for (i = 0; i < TotalNumberToMove; i++)
+        if (SendList[i]->ReturnType() == type) NumberToSend++;
+
+      int *nCount = new int[NumberOfProcessors];
+
+      MPI_Datatype DataTypeInt = (sizeof(int) == 4) ? MPI_INT : MPI_LONG_LONG_INT;
+
+      MPI_Allgather(&NumberToSend, 1, DataTypeInt,
+		    nCount, 1, DataTypeInt, MPI_COMM_WORLD);
+
+      NumberOfNewParticles = 0;
+      for (i = 0; i < NumberOfProcessors; i++) {
+	NumberOfNewParticles += nCount[i];
+      }
+
+      /* Create a MPI packed buffer from the SendList */
+      
+      instance_size = ap_info->ReturnElementSize();
+
+      if (NumberToSend != 0)
+	local_buffer_size = NumberToSend*instance_size;
+      else
+	local_buffer_size = 0;
+
+      int send_buffer_size;
+      char *send_buffer = new char[local_buffer_size];
+      // This will break if more than one AP type is in the simulation
+      send_buffer_size = ap_info->FillBuffer(SendList, NumberToSend, send_buffer);
```

Comment:
It does break for multi-type AP in the simulation. Needs to be fixed before merging.

Reply:
Indeed this is not good. I think an issue needs to be raised on this.

Reply:
I've never tested the code with multiple APs. I suspect many parts of the code would break.

Reply:
Created issue #76